File: src-backend/routes/campaigns.js
const express = require('express');
const { supabase } = require('../db/client');
const { sendPendingBatch } = require('../services/pacing');

const router = express.Router();

// 1. Crear una campaña en modo borrador
router.post('/', async (req, res) => {
  const { name, templateId, meta_template_id } = req.body;
  try {
    const { data: campaign, error } = await supabase
      .from('campaigns')
      .insert([
        {
          name,
          template_id: templateId,
          meta_template_id: meta_template_id || null,
          status: 'draft'
        }
      ])
      .select()
      .single();

    if (error) throw error;

    res.json({ campaignId: campaign.id });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

// 2. Programar y poner en cola los mensajes de una campaña (Inserción masiva optimizada)
router.post('/:id/schedule', async (req, res) => {
  const { id } = req.params;
  const { messageBody, contactIds = [] } = req.body;

  try {
    // Verificar si la campaña existe
    const { data: campaign, error: campaignError } = await supabase
      .from('campaigns')
      .select('*')
      .eq('id', id)
      .maybeSingle();

    if (campaignError) throw campaignError;
    if (!campaign) {
      return res.status(404).json({ error: 'Campaña no encontrada' });
    }

    if (!contactIds.length) {
      return res.status(400).json({ error: 'Debe seleccionar al menos un contacto' });
    }

    // Actualizar el estado de la campaña a 'running' y registrar fecha
    const { error: updateError } = await supabase
      .from('campaigns')
      .update({ status: 'running', scheduled_at: new Date().toISOString() })
      .eq('id', id);

    if (updateError) throw updateError;

    // Obtener los datos de todos los contactos seleccionados (.in maneja la lista de IDs de forma nativa)
    const { data: contacts, error: contactsError } = await supabase
      .from('contacts')
      .select('id, phone, var1, var2')
      .in('id', contactIds);

    if (contactsError) throw contactsError;

    // Preparar el bloque de mensajes para hacer una inserción masiva (Bulk Insert)
    const messagesToInsert = contacts.map(contact => {
      const vars = { var1: contact.var1 || '', var2: contact.var2 || '' };
      return {
        campaign_id: id,
        contact_id: contact.id,
        phone: contact.phone,
        body: messageBody,
        vars_json: vars, // Supabase/PostgreSQL procesa el objeto JSON nativamente sin necesidad de JSON.stringify
        status: 'pending'
      };
    });

    // Insertar todos los mensajes en una sola consulta
    const { error: insertMessagesError } = await supabase
      .from('messages')
      .insert(messagesToInsert);

    if (insertMessagesError) throw insertMessagesError;

    res.json({ ok: true, queued: contacts.length });
    
    // Ejecutar el procesador en segundo plano
    sendPendingBatch();
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

// 3. Obtener métricas y estado actual de los mensajes de una campaña
router.get('/:id/status', async (req, res) => {
  const { id } = req.params;
  try {
    // Agrupamos usando funciones de selección nativas de Supabase
    const { data: statusCounts, error } = await supabase
      .from('messages')
      .select('status')
      .eq('campaign_id', id);

    if (error) throw error;

    // Reducimos los datos para entregar exactamente el formato agrupado que el frontend requiere
    const counts = statusCounts.reduce((acc, current) => {
      const existing = acc.find(item => item.status === current.status);
      if (existing) {
        existing.count++;
      } else {
        acc.push({ status: current.status, count: 1 });
      }
      return acc;
    }, []);

    res.json({ status: counts });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

module.exports = router;

File: src-backend/routes/campaigns.py
import json
import asyncio
from fastapi import APIRouter, HTTPException, status
from pydantic import BaseModel
from typing import List, Optional, Dict, Any
from db.client import db_client

router = APIRouter(prefix="/api/campaigns", tags=["campaigns"])

class CreateCampaignSchema(BaseModel):
    name: str
    templateId: str
    meta_template_id: Optional[str] = None

class ScheduleCampaignSchema(BaseModel):
    messageBody: str
    contactIds: List[int]

@router.post("")
async def create_campaign_endpoint(payload: CreateCampaignSchema):
    """
    Crea un borrador de campaña publicitaria.
    """
    try:
        sql = """
            INSERT INTO campaigns (name, template_id, meta_template_id, status, created_at)
            VALUES (?, ?, ?, 'draft', datetime('now'))
        """
        result = await db_client.run(sql, (payload.name, payload.templateId, payload.meta_template_id))
        return {"campaignId": result["lastID"]}
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))

@router.post("/{campaign_id}/schedule")
async def schedule_campaign_endpoint(campaign_id: int, payload: ScheduleCampaignSchema):
    """
    Programa y encola los mensajes masivos de la campaña para envío con pacing.
    """
    try:
        # 1. Validar la campaña
        campaign = await db_client.get("SELECT * FROM campaigns WHERE id = ?", (campaign_id,))
        if not campaign:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Campaña no encontrada")
            
        if not payload.contactIds:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Debe seleccionar al menos un contacto")
            
        # 2. Actualizar estado de campaña a en curso (running)
        await db_client.run(
            "UPDATE campaigns SET status = 'running', scheduled_at = datetime('now') WHERE id = ?", 
            (campaign_id,)
        )
        
        # 3. Obtener contactos seleccionados
        placeholders = ",".join("?" for _ in payload.contactIds)
        contacts_query = f"SELECT id, phone, var1, var2 FROM contacts WHERE id IN ({placeholders})"
        contacts = await db_client.all(contacts_query, tuple(payload.contactIds))
        
        # 4. Insertar mensajes pendientes en lote
        insert_msg_sql = """
            INSERT INTO messages (campaign_id, contact_id, phone, body, vars_json, status, created_at)
            VALUES (?, ?, ?, ?, ?, 'pending', datetime('now'))
        """
        
        batch_params = []
        for contact in contacts:
            vars_dict = {
                "var1": contact.get("var1") or "",
                "var2": contact.get("var2") or ""
            }
            batch_params.append((
                campaign_id,
                contact["id"],
                contact["phone"],
                payload.messageBody,
                json.dumps(vars_dict)
            ))
            
        if batch_params:
            await db_client.executemany(insert_msg_sql, batch_params)
            
        # 5. Ejecutar procesador de envíos de forma no bloqueante (asyncio task)
        from services.pacing import send_pending_batch
        asyncio.create_task(send_pending_batch())
        
        return {"ok": True, "queued": len(contacts)}
    except HTTPException as he:
        raise he
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))

@router.get("/{campaign_id}/status")
async def get_campaign_status_endpoint(campaign_id: int):
    """
    Retorna métricas de entrega de los mensajes vinculados a la campaña.
    """
    try:
        status_counts = await db_client.all(
            "SELECT status, COUNT(*) as count FROM messages WHERE campaign_id = ? GROUP BY status",
            (campaign_id,)
        )
        return {"status": status_counts}
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))
